File: library/src/main/java/com/pengrad/telegrambot/request/SendPhoto.java
package com.pengrad.telegrambot.request;

import com.pengrad.telegrambot.model.MessageEntity;
import com.pengrad.telegrambot.model.request.ParseMode;

import java.io.File;

/**
 * stas
 * 5/1/16.
 */
public class SendPhoto extends AbstractMultipartRequest<SendPhoto> {

    public SendPhoto(Object chatId, String photo) {
        super(chatId, photo);
    }

    public SendPhoto(Object chatId, File photo) {
        super(chatId, photo);
    }

    public SendPhoto(Object chatId, byte[] photo) {
        super(chatId, photo);
    }

    public SendPhoto caption(String caption) {
        return add("caption", caption);
    }

    public SendPhoto parseMode(ParseMode parseMode) {
        return add("parse_mode", parseMode.name());
    }

    public SendPhoto captionEntities(MessageEntity... entities) {
        return add("caption_entities", entities);
    }

    public SendPhoto hasSpoiler(boolean has_spoiler) {
        return add("has_spoiler", has_spoiler);
    }

    public SendPhoto showCaptionAboveMedia(Boolean showCaptionAboveMedia) {
        return add("show_caption_above_media", showCaptionAboveMedia);
    }

    @Override
    protected String getFileParamName() {
        return "photo";
    }

    @Override
    public String getDefaultFileName() {
        return ContentTypes.PHOTO_FILE_NAME;
    }

    @Override
    public String getDefaultContentType() {
        return ContentTypes.PHOTO_MIME_TYPE;
    }
}

File: library/src/main/java/com/pengrad/telegrambot/request/SendPhoto.kt
package com.pengrad.telegrambot.request

import com.pengrad.telegrambot.model.MessageEntity
import com.pengrad.telegrambot.model.request.ParseMode
import com.pengrad.telegrambot.utility.kotlin.optionalRequestParameter
import java.io.File

class SendPhoto private constructor(
    chatId: Long? = null,
    channelUsername: String? = null,

    photoUrl: String? = null,
    photoFile: File? = null,
    photoBytes: ByteArray? = null
) : KAbstractMultipartRequest<SendPhoto>(
    chatId = chatId,
    channelUsername = channelUsername,

    contentParameterName = "photo",
    contentUrl = photoUrl,
    contentFile = photoFile,
    contentBytes = photoBytes,

    defaultFileName = ContentTypes.PHOTO_FILE_NAME,
    defaultContentType = ContentTypes.PHOTO_MIME_TYPE
) {

    constructor(chatId: Long, photoUrl: String) : this(
        chatId = chatId,
        channelUsername = null,
        photoUrl = photoUrl
    )

    constructor(channelUsername: String, photoUrl: String) : this(
        chatId = null,
        channelUsername = channelUsername,
        photoUrl = photoUrl
    )


    constructor(chatId: Long, photoFile: File) : this(
        chatId = chatId,
        channelUsername = null,
        photoFile = photoFile
    )

    constructor(channelUsername: String, photoFile: File) : this(
        chatId = null,
        channelUsername = channelUsername,
        photoFile = photoFile
    )


    constructor(chatId: Long, photoBytes: ByteArray) : this(
        chatId = chatId,
        channelUsername = null,
        photoBytes = photoBytes
    )

    constructor(channelUsername: String, photoBytes: ByteArray) : this(
        chatId = null,
        channelUsername = channelUsername,
        photoBytes = photoBytes
    )
    

    @Deprecated("Use constructor with chatId or channelUsername instead", ReplaceWith("SendPhoto(chatId, photo)"))
    constructor(chatId: Any, photo: String) : this(
        chatId = (chatId as? Number)?.toLong(),
        channelUsername = chatId as? String,
        photoUrl = photo
    ) {
        checkDeprecatedConstructorParameters()
    }

    @Deprecated("Use constructor with chatId or channelUsername instead", ReplaceWith("SendPhoto(chatId, photo)"))
    constructor(chatId: Any, photo: File) : this(
        chatId = (chatId as? Number)?.toLong(),
        channelUsername = chatId as? String,
        photoFile = photo
    ) {
        checkDeprecatedConstructorParameters()
    }

    @Deprecated("Use constructor with chatId or channelUsername instead", ReplaceWith("SendPhoto(chatId, photo)"))
    constructor(chatId: Any, photo: ByteArray) : this(
        chatId = (chatId as? Number)?.toLong(),
        channelUsername = chatId as? String,
        photoBytes = photo
    ) {
        checkDeprecatedConstructorParameters()
    }

    @Suppress("NOTHING_TO_INLINE")
    private inline fun checkDeprecatedConstructorParameters() {
        if (this.chatId == null && this.channelUsername == null) {
            throw IllegalArgumentException("chatId parameter must be either Long or String")
        }
    }

    var caption: String? by optionalRequestParameter()
    var parseMode: ParseMode? by optionalRequestParameter()
    var captionEntities: List<MessageEntity>? by optionalRequestParameter()
    var showCaptionAboveMedia: Boolean? by optionalRequestParameter()
    var hasSpoiler: Boolean? by optionalRequestParameter()

    fun caption(caption: String) = applySelf { this.caption = caption }

    fun parseMode(parseMode: ParseMode) = applySelf { this.parseMode = parseMode }

    fun captionEntities(captionEntities: List<MessageEntity>) = applySelf { this.captionEntities = captionEntities }

    fun captionEntities(vararg captionEntities: MessageEntity) = applySelf { this.captionEntities = captionEntities.toList() }

    fun showCaptionAboveMedia(showCaptionAboveMedia: Boolean) = applySelf { this.showCaptionAboveMedia = showCaptionAboveMedia }

    fun hasSpoiler(hasSpoiler: Boolean) = applySelf { this.hasSpoiler = hasSpoiler }

}